scanpy/plotting/_tools/scatterplots.py

```diff
@@ -369,26 +378,36 @@ def embedding(
                         rasterized=settings._vector_friendly,
                         **kwargs,
                     )
-                cax = ax.scatter(
-                    _data_points[in_groups, 0],
-                    _data_points[in_groups, 1],
-                    s=in_groups_size,
-                    marker=".",
-                    c=color_vector[in_groups],
-                    rasterized=settings._vector_friendly,
-                    **kwargs,
-                )
+                    cax = ax.scatter(
+                        _data_points[in_groups, 0],
+                        _data_points[in_groups, 1],
+                        s=in_groups_size,
+                        marker=".",
+                        c=color_vector[in_groups],
+                        rasterized=settings._vector_friendly,
+                        **kwargs,
+                    )
 
             else:
-                cax = ax.scatter(
+                if img_key is not None:
+                    cax = circles(
                     _data_points[:, 0],
                     _data_points[:, 1],
-                    s=size,
-                    marker=".",
+                    s=spot_size,
                     c=color_vector,
                     rasterized=settings._vector_friendly,
                     **kwargs,
-                )
+                    )
+                else:
+                    cax = ax.scatter(
+                        _data_points[:, 0],
+                        _data_points[:, 1],
+                        s=size,
+                        marker=".",
+                        c=color_vector,
+                        rasterized=settings._vector_friendly,
+                        **kwargs,
+                    )
```

Comment:
Please just do once at the top:

```py
scatter = ax.scatter if img_key is None else partial(circles, ax=ax)
```

Make sure circles uses the same arguments and order as `scatter`.

Then you can just replace `cax = ax.scatter(` with `cax = scatter(` everywhere instead of building even more huge if blocks.